modules/auxiliary/scanner/rsync/modules_list.rb

```diff
@@ -6,64 +6,223 @@
+  def rsync_parse_lines(response_data)
+    control_lines = []
+    data_lines = []
+
+    if response_data
+      response_data.strip!
+      response_data.split(/\n/).map do |line|
+        if line =~ /^#{RSYNC_HEADER}/
+          control_lines << line
+        else
+          data_lines << line
+        end
+      end
+    end
+
+    [ control_lines, data_lines ]
   end
 
   def run_host(ip)
-    connect
-    version = sock.get_once
-
-    return if version.blank?
-
-    print_good("#{ip}:#{rport} - rsync #{version.strip} found")
-    report_service(:host => ip, :port => rport, :proto => 'tcp', :name => 'rsync')
-    report_note(
-        :host => ip,
-        :proto => 'tcp',
-        :port => rport,
-        :type => 'rsync_version',
-        :data => version.strip
-    )
+    begin
+      connect
+      version, motd = rsync_negotiate
+      unless version
+        vprint_error("#{peer} - does not appear to be rsync")
+        disconnect
+        return
+      end
+    rescue => e
+      vprint_error("#{peer} - error while connecting and negotiating: #{e}")
+      disconnect
+      return
+    end
 
-    # making sure we match the version of the server
-    sock.puts("#{version}")
-    # the listing command
-    sock.puts("\n")
-    listing = sock.get(20)
-    disconnect
-
-    return if listing.blank?
-
-    print_good("#{ip}:#{rport} - rsync listing found")
-    listing.gsub!('@RSYNCD: EXIT', '') # not interested in EXIT message
-    listing_sanitized = Rex::Text.to_hex_ascii(listing.strip)
-
-    vprint_status("#{ip}:#{rport} - #{version.rstrip} #{listing_sanitized}")
-    report_note(
-        :host => ip,
-        :proto => 'tcp',
-        :port => rport,
-        :type => 'rsync_listing',
-        :data => listing_sanitized
+    info = "rsync protocol version #{version}"
+    info += ", MOTD '#{motd}'" if motd
+    report_service(
+      host: ip,
+      port: rport,
+      proto: 'tcp',
+      name: 'rsync',
+      info: info
     )
+    print_status("#{peer} - rsync version: #{version}") if datastore['SHOW_VERSION']
+    print_status("#{peer} - rsync MOTD: #{motd}") if motd && datastore['SHOW_MOTD']
+
+    modules_metadata = {}
+    begin
+      modules_metadata = rsync_list
+    rescue => e
+      vprint_error("#{peer} -- error while listing modules: #{e}")
+      return
+    ensure
+      disconnect
+    end
+
+    if modules_metadata.empty?
+      print_status("#{peer} - no rsync modules found")
+    else
+      modules = modules_metadata.map { |m| m[:name] }
+      print_good("#{peer} - #{modules.size} rsync modules found: #{modules.join(', ')}")
+
+      table_columns = %w(Name Comment)
+      if datastore['TEST_AUTHENTICATION']
+        table_columns << 'Authentication'
+        modules_metadata.each do |module_metadata|
+          begin
+            connect
+            rsync_negotiate
+            module_metadata[:authentication] = get_rsync_auth_status(module_metadata[:name])
+          rescue => e
```

Comment:
Part of this is a side effect of the rsync protocol; discovering what modules are available requires a connect/disconnect, and testing each module each require a separate connect/disconnect.  IOW, there are lots of connects/disconnects that need to happen regardless of whether or not exceptions are encountered.

I put the rescues in to give the user some idea of when things go wrong, but as you point out, catching all exceptions has negative side effects.

I'll see if I can clean this up.